cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,315 @@
+				logger.Warnln(err)
+				continue
+			}
+			return err
+		}
+	}
+	return fmt.Errorf("no valid plugin executable file was found in %s. Plugin executable file name must in format notation-{plugin-name}", zipPath)
+}
+
+// installPluginFromTarGz extracts and untar a plugin tar.gz file, validates and
+// installs the plugin
+func installPluginFromTarGz(ctx context.Context, tarGzPath string, force bool) error {
+	logger := log.GetLogger(ctx)
+	r, err := os.Open(tarGzPath)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	decompressedStream, err := gzip.NewReader(r)
+	if err != nil {
+		return err
+	}
+	defer decompressedStream.Close()
+	tarReader := tar.NewReader(decompressedStream)
+	for {
+		header, err := tarReader.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if strings.Contains(header.Name, "..") {
+			continue
+		}
+		fmode := header.FileInfo().Mode()
+		// requires one and only one executable file, with name in format
+		// notation-{plugin-name}, exists in the tar.gz file
+		if fmode.IsRegular() && osutil.IsOwnerExecutalbeFile(fmode) {
+			err := installPluginExecutable(ctx, header.Name, tarReader, fmode, force)
+			if errors.As(err, &notationerrors.ErrorInvalidPluginName{}) {
+				logger.Warnln(err)
+				continue
+			}
+			return err
+		}
+	}
+	return fmt.Errorf("no valid plugin executable file was found in %s. Plugin executable file name must in format notation-{plugin-name}", tarGzPath)
+}
+
+// installPluginExecutable extracts, validates, and installs a plugin from
+// reader
+func installPluginExecutable(ctx context.Context, fileName string, fileReader io.Reader, fmode fs.FileMode, force bool) error {
+	// sanity check
+	pluginName, err := notationplugin.ExtractPluginNameFromExecutableFileName(fileName)
+	if err != nil {
+		return err
+	}
+	if runtime.GOOS == "windows" && filepath.Ext(fileName) != ".exe" {
+		return fmt.Errorf("on Windows, plugin executable file name %s is missing the '.exe' extension", fileName)
+	}
+	if runtime.GOOS != "windows" && filepath.Ext(fileName) == ".exe" {
+		return fmt.Errorf("on %s, plugin executable file name %s cannot have the '.exe' extension", runtime.GOOS, fileName)
+	}
+	// check plugin existence
+	if !force {
+		existed, err := notationplugin.CheckPluginExistence(ctx, pluginName)
+		if err != nil {
+			return fmt.Errorf("failed to check plugin existence, %w", err)
+		}
+		if existed {
+			return fmt.Errorf("plugin %s already installed", pluginName)
+		}
+	}
+	// extract to tmp dir
+	tmpDir, err := os.MkdirTemp("", notationPluginTmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to create notationPluginTmpDir, %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpFilePath := filepath.Join(tmpDir, fileName)
+	pluginFile, err := os.OpenFile(tmpFilePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fmode)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(pluginFile, fileReader); err != nil {
```

Comment:
`pluginFile` is never closed if `io.Copy` fails. Due to file lock, line 281 will fail silently.

Reply:
Code updated.